### ISSUE TITLE
predict method

### ISSUE THREAD

**Issue by matudor** [kno10/python-kmedoids#22]
Hi, thanks for the nice package.
if my reading of the sklearn API predict method is correct,
https://github.com/kno10/python-kmedoids/blob/125f45f152a3f2f46bdc3c1d405effb56eb98028/kmedoids/__init__.py#L853
it assumes that the test data is the same as the training data, i.e. has the medoids at the same indices.  If so, this precludes application to new datasets, e.g. as one would do in context of CLARA where subset is used to define medoids and then larger set is clustered to nearest medoids.
Without having stepped through the entire codebase, I'm not certain: Is this desirable/necessary behavior, or can it be relaxed to allow new data?

**Comment by kno10** [comment]
The problem is that k-medoids is supposed to be used with a *pairwise* distance matrix.
The other metrics are translated to this case; in an inefficient manner.
The clusterer does not keep track of its "training" data, so you need to pass this again.

The current predict function only makes sense if you extend the distance matrix (or data matrix) with additional samples, keeping all existing objects in place - but you do not need the distances to anything except the medoids.
So `fit(X); predict(X)` returns the same as `fit_predict(X)`.

Because sklearn lacks sufficient type safety (distance matrix vs. data matrix!) people are very likely to use this API incorrectly.

In my opinion, the `predict` API of scikit-learn is badly designed. It makes sense only for k-means - and for classification.

@lale1009 we should probably just implement this with an error.
```python
raise NotImplmentedError("This API is not safe to use. Use the argmin of the distances to the medoids.")
```
maybe along with a shortcut to return the labels if the X is the same that was used when fitting (although I'd rather not want to keep a reference to it - does Python have anything such as a weak reference?)
Or maybe we can support the non-`precomputed` case in the sklearn wrapper, keeping a reference to the medoid data rows for other metrics?

**Comment by lale1009** [comment]
Supporting the non-precomputed case and raise exception for precomputed case with coming:
d97c99bf156879ba718038b2191ae785f0a1bbd7
